### ISSUE TITLE
Feature Request: Cloudfront Origin Failover

### ISSUE THREAD

**Issue by danieladams456** [terraform-providers/terraform-provider-aws#6547]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
Cloudfront now allows origin groups to manage failover if one origin is unavailable.  The group contains origin IDs and status codes under which to fail over.

This feature will be very useful for increasing availability of static sites hosted on S3 + Cloudfront without needing lambda@edge hacks.

### New or Affected Resource(s)
- aws_cloudfront_distribution

### Potential Terraform Configuration
```hcl
resource "aws_cloudfront_distribution" "s3_distribution" {
  origin_group {
    origin_id = "groupS3"

    failover_criteria {
      status_codes = [403, 404, 500, 502, 503, 504]
    }

    members {
      # see note below
      ordered_origin_group_member {
        origin_id = "primaryS3"
      }
      ordered_origin_group_member {
        origin_id = "failoverS3"
      }
    }
  }

  origin {
    domain_name = "${aws_s3_bucket.primary.bucket_domain_name}"
    origin_id   = "primaryS3"

    s3_origin_config {
      origin_access_identity = "${aws_cloudfront_origin_access_identity.default.cloudfront_access_identity_path}"
    }
  }

  origin {
    domain_name = "${aws_s3_bucket.failover.bucket_domain_name}"
    origin_id   = "failoverS3"

    s3_origin_config {
      origin_access_identity = "${aws_cloudfront_origin_access_identity.default.cloudfront_access_identity_path}"
    }
  }

  default_cache_behavior {
    target_origin_id       = "groupS3"
  }
}
```
It would probably align more to the Cloudfront API data types to do `ordered_origin_group_member` blocks inside the `members` block.  The [XML API](https://docs.aws.amazon.com/cloudfront/latest/APIReference/API_OriginGroupMembers.html) uses a list of exactly two items that appear to respect the item order for priority.  Aliasing them to `primary_origin_id` and `failover_origin_id` would be more friendly, but might require a breaking change later if the allowed number changed.

### References
- [AWS Announcement](https://aws.amazon.com/about-aws/whats-new/2018/11/amazon-cloudfront-announces-support-for-origin-failover/)
- [AWS Docs](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/high_availability_origin_failover.html)
- [OriginGroupMembers API Data Type](https://docs.aws.amazon.com/cloudfront/latest/APIReference/API_OriginGroupMembers.html)

**Comment by jrstarke** [comment]
I'm interested in this functionality as well.

Looks like the [DistributionConfig](https://docs.aws.amazon.com/sdk-for-go/api/service/cloudfront/#DistributionConfig) of the AWS GO SDK hasn't been updated to include the new [OriginGroups](https://docs.aws.amazon.com/cloudfront/latest/APIReference/API_DistributionConfig.html#cloudfront-Type-DistributionConfig-OriginGroups) element of the [DistributionConfig](https://docs.aws.amazon.com/cloudfront/latest/APIReference/API_DistributionConfig.html#cloudfront-Type-DistributionConfig-OriginGroups) in the AWS API yet.

**Comment by bflad** [comment]
FYI, it is updated in AWS Go SDK v1.15.80 (and the AWS Go SDK dependency updated in this codebase has been bumped to v1.15.81): https://github.com/terraform-providers/terraform-provider-aws/pull/6572/files#diff-315b52f0aed6c83c83f0642081d29c08R8354

Sometimes the AWS Go SDK API Reference publisher has a delay (I've seen it take up to a day or two, but never this long). It might be related to re:Invent this week. Anyways, I created this upstream issue: https://github.com/aws/aws-sdk-go/issues/2301

**Comment by balassit** [comment]
@bflad is there any update on this? I saw the docs updated. It would be great to get this behavior introduced to terraform

**Comment by bflad** [comment]
The only update I can provide as a maintainer is that this is not being actively worked on by any of the HashiCorp maintainers (that I know about) in the near future. From a community perspective, looks like this is working its way up the [community issue voting](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc), so at a certain point if its not handled with a community contribution, it might be put on our internal roadmap. Keep those 👍 coming (on the original issue above) if this is something you want.

**Comment by bflad** [comment]
Support for `aws_cloudfront_distribution` resource `origin_group` configuration blocks has been merged with an example in resource documentation, e.g.

```hcl
resource "aws_cloudfront_distribution" "s3_distribution" {
  # ... other configuration ...

  origin {
    domain_name = "${aws_s3_bucket.primary.bucket_regional_domain_name}"
    origin_id   = "primaryS3"

    s3_origin_config {
      origin_access_identity = "${aws_cloudfront_origin_access_identity.default.cloudfront_access_identity_path}"
    }
  }

  origin {
    domain_name = "${aws_s3_bucket.failover.bucket_regional_domain_name}"
    origin_id   = "failoverS3"

    s3_origin_config {
      origin_access_identity = "${aws_cloudfront_origin_access_identity.default.cloudfront_access_identity_path}"
    }
  }

  origin_group {
    origin_id = "groupS3"

    failover_criteria {
      status_codes = [403, 404, 500, 502]
    }

    member {
      origin_id = "primaryS3"
    }

    member {
      origin_id = "failoverS3"
    }
  }

  default_cache_behavior {
    # ... other configuration ...

    target_origin_id = "groupS3"
  }
}
```

This will release with version 2.3.0 of the Terraform AWS Provider, likely middle of this week. 👍 For any future feature requests or bug reports, please open a new GitHub issue.

**Comment by bflad** [comment]
This has been released in [version 2.3.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#230-march-21-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
